### PR TITLE
Enable 32-bit lowering of tosa.apply_scale for RISC-V

### PR DESCRIPTION
**Pull request by dcaballe**
This patch enables the 32-bit lowering of 'tosa.apply_scale' when
we target RISC-V without 64-bit integer vector support (i.e.,
without +v and/or +zve64x features).

### CODE REVIEW

**Comment by benvanik** [comment]
Drive by: if this is correct can we flip the condition so this is always used _unless_ the optimized version is available?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+// -----
+
+#executable_layout = #hal.executable.layout<push_constants = 0, sets = [
+  #hal.descriptor_set.layout<0, bindings = [
+    #hal.descriptor_set.binding<0, storage_buffer>,
+    #hal.descriptor_set.binding<1, storage_buffer>
+  ]>
+]>
+#executable_target_embedded_elf_riscv_64_ = #hal.executable.target<"llvm", "embedded-elf-riscv_64", {
+  cpu_features = "+m,+a,+f,+d,+c,+zve32x,+zve64x",
```
Comment (hcindyl) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir:102]: Should we add both zve flags? zve64x implies zve32x

https://github.com/llvm/llvm-project/tree/232c00b/llvm/lib/Support/RISCVISAInfo.cpp#L754

Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir:102]: Good point! Let me remove one. Thanks!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        %2 = vector.load %0[%c0] : memref<2xi32>, vector<2xi32>
+        %3 = "tosa.apply_scale"(%2, %cst, %cst_0) {double_round = false} : (vector<2xi32>, vector<2xi32>, vector<2xi8>) -> vector<2xi32>
+        vector.store %3, %1[%c0] : memref<2xi32>, vector<2xi32>
+        return
+      }
+    }
+  }
+}
+
+// 32-bit lowering should be used by default if no vector features are provided.
```
Comment (hcindyl) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir:41]: 64-bit?

**Comment by dcaballe** [comment]
> Drive by: if this is correct can we flip the condition so this is always used _unless_ the optimized version is available?

Not sure I follow. The 64-bit implementation should be better (I'll benchmark this eventually to be sure) if the target implementation has support for 64-bit integer vectors (far less instructions). Otherwise, we should use the 32-bit implementation, which would generate more instructions but avoid 64-bit vectors and, therefore, avoid the scalarization of 64-bit vectors in the backend. What would you like to flip?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir (hunk 2 above)
Comment (hcindyl) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir:41]: oh, I see the logic now. So the intention is not to unroll the scalar-only target, even if the scalar ISA supports int64 (as the target_triple is riscv64 prefixed)?

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp [hunk 3]
```diff
@@ -100,21 +100,63 @@ bool isAArch64(IREE::HAL::ExecutableVariantOp variantOp) {
   return triple && triple.getValue().isAArch64();
 }
 
+static Optional<IREE::HAL::ExecutableVariantOp> getParentExecutableVariantOp(
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:103]: Isnt this same as `getEntryPoint`?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir (hunk 2 above)
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir:41]: Honestly, I didn't think too much about this case since we would always expect at least some flavor of vector extensions... (otherwise, we shouldn't even vectorized the code!). We can change it to the 64-bit version if you think it's more convenient.

**Comment by benvanik** [comment]
Ah sorry - I may not know what this is doing then; I'd assumed this was using 32-bit math instead of 64-bit given the PR description, but if this is just doing the same 64-bit math using 32-bit registers for emulation then that's different. It's unclear from the lit tests because they all look to be doing 64-bit math - I'm not actually sure how to read the changes. Not important I understand, but maybe clarify the description or the test comments ("32-bit lowering should be used" still seems to use 64bit, etc)?

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp (hunk 3 above)
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:103]: Ah, I missed that one! They seem different. This one is returning an `ExecutableVarianOp`, which is the one with the target information. `getEntryPoint` returns an `ExecutableExportOp`. I think we should keep both but I can align the new one with the implementation with `getEntryPoint`.

**Comment by dcaballe** [comment]
> Ah sorry - I may not know what this is doing then; I'd assumed this was using 32-bit math instead of 64-bit given the PR description, but if this is just doing the same 64-bit math using 32-bit registers for emulation then that's different. It's unclear from the lit tests because they all look to be doing 64-bit math - I'm not actually sure how to read the changes. Not important I understand, but maybe clarify the description or the test comments ("32-bit lowering should be used" still seems to use 64bit, etc)?

Exactly, we are using 32-bit instructions to emulate the 64-bit computation but it's more complicated than that :). The 32-bit emulation generates a 64-bit matmul and then extract the 32-bit upper and lower parts. Then, the RISC-V backend maps this to 32-bit mul_hi and mul_lo instructions and removes the 64-bit multiplication. That's the best we can do for now :). I'll improve the doc. Thanks!

> (if this was actually using 32-bit math then it'd be better even if the target supports 64-bit; doubling the number of usable bits per cycle is pretty good :)

Agreed!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir (hunk 2 above)
Comment (hcindyl) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir:41]: If we don't expect any vectorization in this case (no +v, no +zve), shouldn't we check for the pattern of the non-vectorized result? Not sure that is something checkable at the LLVM IR level though.

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp [hunk 4]
```diff
@@ -100,21 +100,63 @@ bool isAArch64(IREE::HAL::ExecutableVariantOp variantOp) {
...
+
+/// Returns true if the 'op' is or is enclosed in a HAL::ExecutableVarianOp that
+/// contains '+zve64x' in its cpu features.
+bool hasZve64xFeature(Operation *op) { return hasFeature(op, "+zve64x"); }
+
+bool isRISCV(Operation *op) {
+  auto variantOp = getParentExecutableVariantOp(op);
+  if (!variantOp) {
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:157]: +1. Similar to the comment above. I'd refactor these `has*` and `is*` methods to take the `executableVariantOp` as the operand. The caller can raise an error if there is no `executableVariantOp` found.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Just a follow up comment about the `has*` and `is*` methods added here.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir (hunk 2 above)
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir:41]: Sorry, my comment was misleading! At this point, we assume that we'll always have some kind of vector support so we always (try to) vectorize the code without actually looking at any target features. The first thing we should do if we wanted to support targets without any vector support is to prevent vectorization to happen, which we don't do at this point. That's why what we do for this case (no +v, no +zve) is not relevant in practice. I just added it to test that corner case in the lowering logic. Hopefully that makes more sense :)

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp (hunk 3 above)
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/Utils/Utils.cpp:103]: For this particular case, yes. Sure, I'll align the implementations.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir (hunk 2 above)
Comment (hcindyl) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir:41]: Unfortunately that's not how we state in https://google.github.io/iree/building-from-source/riscv/. We probably should not expect everyone who uses IREE + RISCV backend has a v-ext enabled device.

If we currently can't prevent vectorizing for any target that' can't support vectorizing effectively, IMHO, we should create an issue and resolve it, and we probably should add a TODO comment in this test.

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp (hunk 3 above)
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/Utils/Utils.cpp]: I used `FailureOf<...>` but it feels a bit weird (perhaps it's just me) to check `succeeded(variableOp)` to know if `variableOp` actually has a value. Perhaps we should use `Optional` instead but I won't change it as part of this PR since changes would span to other files.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp [hunk 5]
```diff
@@ -959,7 +959,15 @@ void ConvertToLLVMPass::runOnOperation() {
   //   multiply or add.
   //
   // TODO(bjacob): Use a lowering that uses specific ARM/X86 intrinsics.
-  tosa::populateTosaRescaleToArithConversionPatterns(&patterns);
+  bool use32BitImpl = false;
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp:962]: Based on the comment right above this code, I decided to change the policy and always use the 64-bit lowering by default. That will preserve the existing behavior.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        %3 = "tosa.apply_scale"(%2, %cst, %cst_0) {double_round = false} : (vector<2xi32>, vector<2xi32>, vector<2xi8>) -> vector<2xi32>
+        vector.store %3, %1[%c0] : memref<2xi32>, vector<2xi32>
+        return
+      }
+    }
+  }
+}
+
+// 64-bit lowering is used by default if no vector features are provided.
+// TODO: We shouldn't vectorize the code if no vector features are provided.
```
Comment (hcindyl) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir:42]: nit: TODO(handle) or TODO(issue number)

Do we have an issue for this feature improvement?

**Review by MaheshRavishankar — APPROVED** [review]

**Review by hanhanW — APPROVED** [review]
LGTM

File: compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp [hunk 7]
```diff
@@ -959,7 +959,15 @@ void ConvertToLLVMPass::runOnOperation() {
...
+  auto variantOp = getExecutableVariantOp(module);
+  if (succeeded(variantOp) && isRISCV(*variantOp)) {
+    // Use the 32-bit lowering for RISC-V if 'zve32x' is specified and there is
+    // no 64-bit integer vector support.
+    use32BitImpl = hasZve32xFeature(*variantOp) && !hasVFeature(*variantOp) &&
+                   !hasZve64xFeature(*variantOp);
```
[the remark this answers is not in the document]
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp:968]: I'm trying to be defensive here because unfortunately we don't have a way to know if the features provided are correct or their equivalences. Consequently, we shouldn't be using the 32-bit lowering if `+zve32x, +v` or `+zve32x, +zve64x` are used even though they shouldn't (I actually used `+zve32x, +zve64x` in my experiments and not even LLVM complained :)).

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir (hunk 6 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir:42]: `TODO:` is LLVM style. The IREE style is using `TODO(handle)` or `TODO(#ISSUE_NUMBER)`.

Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/apply_scale_lowering.mlir]: Fixed the TODO, thanks! No, we don't have an issue yet. Feel free to open one with the specific information of your use case so that we can discuss it in more detail.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (hcindyl) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp:968]: I see. It depends on when the check is done. As shown in the LLVM snippet, `+v` implies `+zve64d` and `+zve32x,+v` would be effectively `+zve64d` after https://github.com/llvm/llvm-project/blob/232c00b/llvm/lib/Support/RISCVISAInfo.cpp#L814, but it may not be the case if the check is on the input `--cpu-features` string.

**Review by hcindyl — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp:968]: I just check and got `+m,+a,+f,+d,+c,+v,+zve32x,+zve64x` so it seems we are keeping the raw string around without any processing. I'll open an issue for this. Thanks!
